pkg/authorization/api/types.go

```diff
@@ -0,0 +1,110 @@
+package api
+
+import (
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kruntime "github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	kutil "github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// shouldn't store the information on the users or teams directly because assigning policy is distinct from creating users (or is it?)
+// PolicyRules can be created from a PolicyRuleTemplate plus user/team/account.  PolicyRules can be created from PolicyRuleTemplates
+// in the same names or the global namespace, provided you have rights to read said policies
+// Roles are sets of PolicyRuleTemplates plus the information required to actualize the PolicyRuleTemplates into PolicyRules
+// Authorization is calculated against
+//   1. all allow PolicyRules in the namespace
+//   2. all allow actualized Role PolicyRules in the namespace
+//   3. all allow PolicyRules in the global namespace
+//   4. all allow actualized Role PolicyRules in the global namespace
+//   5. all deny PolicyRules in the namespace
+//   6. all deny actualized Role PolicyRules in the namespace
+//   7. all deny PolicyRules in the global namespace
+//   8. all deny actualized Role PolicyRules in the global namespace
+//   Any allow success short circuits to the denies.  Any deny returns a deny.  Return errors include both the allowing rule and the denying rule.
+
+const (
+	// Policy is a singleton and this is its name
+	PolicyName = "Policy"
+)
+
+// PolicyRule holds information that describes a policy rule, but does not contain information
+// about who the rule applies to or which namespace the rule applies to.
+type PolicyRule struct {
+	// Deny is true if any request matching this rule should be denied.  If false, any request matching this rule is allowed.
+	Deny bool `json:"deny" yaml:"deny"`
+	// Verbs is a list of Verbs that apply to ALL the ResourceKinds and AttributeRestrictions contained in this rule.  "*" represents all kinds.
+	Verbs []string `json:"verbs" yaml:"verbs"`
+	// AttributeRestrictions will vary depending on what the Authorizer/AuthorizationAttributeBuilder pair supports.
+	// If the Authorizer does not recognize how to handle the AttributeRestrictions, the Authorizer should report an error.
+	AttributeRestrictions kruntime.EmbeddedObject `json:"attributeRestrictions" yaml:"attributeRestrictions"`
+	// ResourceKinds is a list of kinds this rule applies to.  "*" represents all kinds.
+	ResourceKinds []string `json:"resourceKinds" yaml:"resourceKinds"`
+}
+
+// Role is a logical grouping of PolicyRules that can be referenced as a unit by RoleBindings.
+type Role struct {
+	kapi.TypeMeta   `json:",inline" yaml:",inline"`
```

Comment:
In other places where we nest objects in objects, do we usually put TypeMeta/ObjectMeta on the nested objects? Just seems like a lot of detail to serialize. Is this needed here because we're exposing the sub-objects as virtual resources?